abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+) -> str:
+    """Collates A/B transformed files into a Parquet dataset.
+
+    This process can be summarized into two (2) important steps:
+        1. Write all transformed JSON records into a temporary Parquet dataset
+           partitioned by the transformed file name.
+        2. For every transformed file, use DuckDB to join A/B Parquet tables
+           using the TIMDEX record ID and write joined records to a Parquet dataset.
+
+    This function (and its subfunctions) uses DuckDB, generators, and batching to
+    write records to Parquet datasets in a memory-efficient manner.
+    """
+    transformed_dataset_path = tempfile.TemporaryDirectory()
+    collated_dataset_path = str(Path(run_directory) / "collated")
+
+    transformed_written_files = write_to_dataset(
+        get_transformed_batches_iter(run_directory, transformed_files),
+        schema=TRANSFORMED_DATASET_SCHEMA,
+        base_dir=transformed_dataset_path.name,
+        partition_columns=["transformed_file_name"],
+    )
+    logger.info(
+        f"Wrote {len(transformed_written_files)} parquet file(s) to transformed dataset"
+    )
+
+    joined_written_files = write_to_dataset(
+        get_joined_batches_iter(transformed_dataset_path.name),
+        base_dir=collated_dataset_path,
+        schema=JOINED_DATASET_SCHEMA,
+    )
+    logger.info(f"Wrote {len(joined_written_files)} parquet file(s) to collated dataset")
+    return collated_dataset_path
+
+
+def yield_records(transformed_file: str | Path) -> Generator:
+    """Yields data for every TIMDEX record in a transformed file.
+
+    This function uses ijson to yield records from a JSON stream
+    (i.e., the transformed file) one at a time. A generator is returned,
+    yielding a dictionary that contains the following:
+
+    * timdex_record_id: The TIMDEX record ID.
+    * source: The shorthand name of the source as denoted in by Transmogrifier
+      (see https://github.com/MITLibraries/transmogrifier/blob/main/transmogrifier/config.py).
+    * record: The TIMDEX record serialized to a JSON string then encoded to bytes.
+    * version: The version of the transform, parsed from the absolute filepath to a
+      transformed file.
+    * transformed_file_name: The name of the transformed file, excluding file extension.
+    """
+    version, transformed_file_name = parse_parquet_details_from_transformed_file(
+        str(transformed_file)
+    )
+    with open(transformed_file) as file:
+        records = ijson.items(file, "item")
+        for record in records:
+            yield {
+                "timdex_record_id": record["timdex_record_id"],
+                "source": record["source"],
+                "record": json.dumps(record).encode(),
+                "version": version,
+                "transformed_file_name": transformed_file_name,
+            }
+
+
+def get_transformed_batches_iter(
+    run_directory: str, transformed_files: tuple[list[str], ...]
+) -> Generator:
```

Comment:
Again, be specific about what the `Generator` is returning in the type hint

Reply:
Updated return type hint to `Generator[pa.RecordBatch]`.